spec.md

```diff
@@ -0,0 +1,954 @@
+  * Container deployment considerations (`CAP_SYS_ADMIN`, mount namespace, etc.).
+
+### Non-Goals in MVP
+
+The Container Machine Interface (CMI) explicitly will not define, provide, or dictate:
+
+* Specific mechanisms by which a Plugin Supervisor manages the lifecycle of a Plugin, including:
+  * How to maintain state (e.g. what is provisioned, ready, etc.).
+  * How to deploy, install, upgrade, uninstall, monitor, or respawn (in case of unexpected termination) Plugins.
+* Protocol-level authentication and authorization.
+* Packaging of a Plugin.
+
+## Solution Overview
+
+This specification defines an interface along with the minimum operational and packaging recommendations for a VM provider to implement a CMI compatible plugin.
+The interface declares the RPCs that a plugin MUST expose: this is the **primary focus** of the CMI specification.
+
+### Architecture
+
+The primary focus of this specification is on the **protocol** between a CMI-Client and a Plugin.
+It SHOULD be possible to ship CMI-Client compatible Plugins for a variety of deployment architectures.
+A CMI-Client SHOULD be equipped to handle both centralized and headless plugins, as well as split-component and unified plugins.
+Several of these possibilities are illustrated in the following figures.
+
+```
+                             MCM Architecture
++-------------------------------------------+
+|                                           |
+|  +------------+           +------------+  |
+|  |   CMI      |   gRPC    |   CMI      |  |
+|  |   Client   +----------->   Plugin   |  |
+|  +------------+           +------------+  |
+|                                           |
++-------------------------------------------+
+
+Figure 1: CMI client and CMI Plugin together make up
+the MCM. Both of them are typically expected to run
+as two containers is a same pod. However, both might
+run in different pods as well.
+```
+
+### Machine Lifecycle
+
+```
+  CreateMachine +------------+ DeleteMachine
+ +------------->|  PENDING   +--------------+
+ |              +---+----+---+              |
+ |               VM |    | VM               v
++++              is |    | is              +++
+|X|              Up |    | Down            | |
++-+             +---v----+---+             +-+
+                | AVAILABLE  |
+                +---+----^---+
+                 VM |    | VM
+        is ready to |    | is not ready to
+      schedule pods |    | schedule pods
+                +---v----+---+
+                |   READY    |
+                +------------+
+
+Figure 2: The lifecycle of a dynamically provisioned machine, from
+creation to destruction.
+```
+
+The above diagrams illustrate a general expectation with respect to how a MCM manages the lifecycle of a machine via the API presented in this specification.
+Plugins SHOULD expose all RPCs for an interface: Controller plugins SHOULD implement all RPCs for the `Machine` service.
```

Comment:
Yes, I agree with you. We need a harder restriction on the required methods and optional ones. Let me try to adapt the proposal to the same.

Reply:
I think should be a minimum set of capabilities that are mandatory since without them, the CMI client couldn't do a meaningful job. I think this should include `CreateMachine` and `DeleteMachine`, not sure about `GetMachine`. The rest should be optional.